src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -769,15 +773,23 @@ internal protected virtual Task OnCloseAsync(TimeSpan timeout)
 #endif
 
             // If we have already executed this code path, just return a completed Task.
-            // It means this type did not override OnCloseAsync or OnBeginClose.
+            // This is a safeguard in case custom CommunicationObjects attempt to do this multiple times.
             if (_onCloseAsyncCalled)
             {
                 return TaskHelpers.CompletedTask();
             }
 
-            // External types cannot override this method because it is not public.
-            // Redirect this call to the APM path.
             _onCloseAsyncCalled = true;
+
+            // External types cannot override this method because it is not public.
+            // If the caller started this open synchronously, we must use the synchronous
+            // methods they may have overridden.
+            if (_isSynchronousClose)
+            {
+                return Task.Run(() => OnClose(timeout));
```

Comment:
Yes, I am also interested in what you propose. This is actually the code that was there before I switched to using OnBeforeOpen. Perhaps there is a way to use AsyncLocal or something like it to detect we are already running in a layered async path?